scripts/task.ts

```diff
@@ -31,166 +58,407 @@ export type TemplateDetails = {
-    junitFilename: junit && getJunitFilename(taskKey),
+function getTaskKey(task: Task): TaskKey {
+  return (Object.entries(tasks) as [TaskKey, Task][]).find(([_, t]) => t === task)[0];
+}
+
+/**
+ *
+ * Get a list of tasks that need to be (possibly) run, in order, to
+ * be able to run `finalTask`.
+ */
+function getTaskList(finalTask: Task, optionValues: PassedOptionValues) {
+  const taskDeps = new Map<Task, Task[]>();
+  // Which tasks depend on a given task
+  const tasksThatDepend = new Map<Task, Task[]>();
+
+  const addTask = (task: Task, dependent?: Task) => {
+    if (tasksThatDepend.has(task)) {
+      if (!dependent) throw new Error('Unexpected task without dependent seen a second time');
+      tasksThatDepend.set(task, tasksThatDepend.get(task).concat(dependent));
+      return;
+    }
+
+    // This is the first time we've seen this task
+    tasksThatDepend.set(task, dependent ? [dependent] : []);
+
+    const beforeTaskNames =
+      typeof task.before === 'function' ? task.before(optionValues) : task.before || [];
+    const beforeTasks = beforeTaskNames.map((n) => tasks[n]);
+    taskDeps.set(task, beforeTasks);
+
+    beforeTasks.forEach((t) => addTask(t, task));
   };
+  addTask(finalTask);
 
-  if (await task.ready(templateKey, details)) {
-    if (mustNotBeReady) throw new Error(`❌ ${taskKey} task has already run, this is unexpected!`);
+  // We need to sort the tasks topologically so we run each task before the tasks that
+  // depend on it. This is Kahn's algorithm :shrug:
+  const sortedTasks = [] as Task[];
+  const tasksWithoutDependencies = [finalTask];
 
-    logger.debug(`✅ ${taskKey} task not required!`);
-    return;
-  }
+  while (taskDeps.size !== sortedTasks.length) {
+    const task = tasksWithoutDependencies.pop();
+    if (!task) throw new Error('Topological sort failed, is there a cyclic task dependency?');
 
-  if (mustBeReady) {
-    throw new Error(`❌ ${taskKey} task has not already run, this is unexpected!`);
+    sortedTasks.unshift(task);
+    taskDeps.get(task).forEach((depTask) => {
+      const remainingTasksThatDepend = tasksThatDepend
+        .get(depTask)
+        .filter((t) => !sortedTasks.includes(t));
+      if (remainingTasksThatDepend.length === 0) tasksWithoutDependencies.push(depTask);
+    });
   }
 
-  if (task.before?.length > 0) {
-    for (const beforeKey of task.before) {
-      await runTask(beforeKey, templateKey, {
-        mustNotBeReady: false,
-        mustBeReady: !before,
-        before,
-        junit: false, // never store junit results for dependent tasks
-      });
-    }
-  }
+  return { sortedTasks, tasksThatDepend };
+}
+
+type TaskStatus =
+  | 'ready'
+  | 'unready'
+  | 'running'
+  | 'complete'
+  | 'failed'
+  | 'serving'
+  | 'notserving';
+const statusToEmoji: Record<TaskStatus, string> = {
+  ready: '🟢',
+  unready: '🟡',
+  running: '🔄',
+  complete: '✅',
+  failed: '❌',
+  serving: '🔊',
+  notserving: '🔇',
```

Comment:
I had no idea what the muted speaker emoji meant when running a task, but seeing this  know a little more. I don't have any better alternatives though.

But I'm not sure this distinction is necessary.
Eg. when running the `dev` flow, `dev` starts as "notserving" but ends up as "running", so it never gets to "serving". Are there any cases where we would actually see "serving"?
I'm considering if we should just stick to unready/ready.

Reply:
If `dev` ends up `running` that's a bug! 👀 

The reason for the distinction is that service tasks don't *have* to end up being executed, it is valid for them to stay `not-serving` if none of their dependencies end up running.

Reply:
If dev ends up running that's a bug! 👀

I guess the dev task doesn't really ever "complete" because it's a service task. Open to improving the UI in various ways! But maybe in another PR :)